### ISSUE TITLE
Difficulty with importing sibling submodules of mutated module

### ISSUE THREAD

**Issue by RunasSudo** [sixty-north/cosmic-ray#198]
Consider the package structure:

* a
  * b
  * c

If package a.b is being mutated (i.e. in the `exec` stage, but not the `init` stage) a subsequent attempt to import a.c will fail with the error that `module 'a' has no attribute 'c'`.

I'm not 100% sure why this happens, but it appears to be due to the fact that the Python meta_path is adjusted, and the new ASTFinder/ASTLoader provided by cosmic-ray is not aware of the existence of package a.c.

Replacing the `del_mods` line and subsequent for loop in importing.py with a simple `del sys.modules[module_name]` seems to fix this issue, but I don't know whether this may introduce other undesirable side effects. A better solution may be possible by making the new meta_path entries aware of the existence of any such sibling packages.

**Comment by abingham** [comment]
I tried to reproduce this with a trivial test that just imports a.b and
a.c, but I don't see any problems. Can you provide a minimal project (i.e.
package and tests) that trigger the problem?

It will not surprise me at all if there are subtle (or even not-so-subtle)
issues with our mutant injection strategy. I picked the simplest, most
obvious approach that seemed correct, and my plan is to improve it as we
find problems like this. So thanks for taking the time to report this!

**Comment by RunasSudo** [comment]
Hmm, I'm also having difficulty replicating this from scratch. The bug is probably more specific that what I outlined above.

I have a project which reliably fails, but it's quite unwieldy, so I've slimmed it down into an MWE: [mwe.zip](https://github.com/sixty-north/cosmic-ray/files/737497/mwe.zip)


After running:

```bash
cosmic-ray init --timeout=10 mwe eos_stjjr.crypto -- cosmic_ray_tests
```

The print statement shows the test completes successfully. But when running:

```bash
cosmic-ray exec mwe # Let this run for a few seconds, then kill it
cosmic-ray report --full-report mwe
```

cosmic-ray reports a survival rate of 0.00%, but each item reports the error:

```
  File "eos_stjjr.crypto", line 22, in <module>
AttributeError: module 'eos_stjjr' has no attribute 'bigint'
```

**Comment by abingham** [comment]
I'm able to reproduce the issue with the project you provided. Hopefully
I'll get this sorted soon. Thanks!

**Comment by grapherd** [comment]
Hi @RunasSudo, after some try, I think this problem is caused by import rules, not CR itself.

1. [mutual top-level import](http://stackoverflow.com/questions/1250103/attributeerror-module-object-has-no-attribute)
After changing code to above, problem solve
```python
# To trigger mutation
condition = ((2 + 2) == 5)
def trigger():
    import eos_stjjr.bigint
    # Triggers the error
    eos_stjjr.bigint.BigInt('42')
```

2. [intra-package reference](https://docs.python.org/3/tutorial/modules.html#intra-package-references)

If we change `import eos_stjjr.bigint` to `from . import bigint`, then the code can change to
```python
condition = ((2 + 2) == 5)

from . import bigint

# Triggers the error
bigint.BigInt('42')
```

Hope this will help you.

**Comment by abingham** [comment]
@grapherd is right that this will make your testing work, but it's not a complete solution. In the end, CR should never force users to change their code; it should do its job correctly on any valid code base (like yours).

I *think* the problem that CR is having is that I didn't fully account for how Python handles references to modules. Before we run a test suite, we attempt to remove any references to the mutated module so that all points in the code will re-import it before they use it. I think my algorithm for this is insufficient, and I'm going to create a branch soon with a (hopefully) improved version.

**Comment by RunasSudo** [comment]
@grapherd Sorry, could you explain how the code represents a mutual top-level import? eos_stjjr.crypto imports eos_stjjr.bigint, but there is no import in the other direction.

The fact that the code runs perfectly fine without cosmic-ray, and indeed, runs perfectly fine *with* cosmic-ray provided that no mutations are attempted, seems to suggest that the problem is not with the code but with cosmic-ray's handling of it.

(The two suggestions may function as workarounds generally, but as mentioned, I believe the problem lies fundamentally with cosmic-ray, and a re-factor should not be necessary to allow cosmic-ray to run otherwise perfectly passable code.)

**Comment by abingham** [comment]
> a re-factor should not be necessary to allow cosmic-ray to run otherwise perfectly passable code

Correct, and I'll try to address this soon.

**Comment by abingham** [comment]
@RunasSudo Try out the [fixing-import-scrubber](https://github.com/sixty-north/cosmic-ray/tree/fixing-import-scrubber) branch. This should address the problem you're seeing.

I'm not sure that this approach is 100% correct, but it's certainly a step in the right direction.

**Comment by grapherd** [comment]
@abingham @RunasSudo sorry for that, is truly CR problem, not others.

[fixing-import-scrubber](https://github.com/sixty-north/cosmic-ray/tree/fixing-import-scrubber) working right at Python 3.6 here.

**Comment by RunasSudo** [comment]
@abingham The [fixing-import-scrubber](https://github.com/sixty-north/cosmic-ray/tree/fixing-import-scrubber) seems to have solved the issue for me.

**Comment by abingham** [comment]
This should be fixed as of ba4d2e662d81657bae1bfe982b6a59edaf2fee98, but let me know if you see any problems. 

This really needs a test to prevent regressions, and I've made a reminder to make on in #200.
